src/Sylius/Component/Core/Promotion/Action/ItemPercentageDiscountAction.php

```diff
@@ -0,0 +1,164 @@
+        }
+
+        $filteredItems = $this->taxonFilter->filter($subject->getItems(), $configuration);
+
+        foreach ($filteredItems as $item) {
+            $promotionAmount = (int) round($item->getTotal() * $configuration['percentage']);
+            $distributedAmounts = $this->distributor->distribute($promotionAmount, $item->getQuantity());
+
+            $this->setUnitsAdjustments($item, $distributedAmounts, $promotion);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function revert(PromotionSubjectInterface $subject, array $configuration, PromotionInterface $promotion)
+    {
+        if (!$subject instanceof OrderInterface) {
+            throw new UnexpectedTypeException($subject, OrderInterface::class);
+        }
+
+        foreach ($subject->getItems() as $item) {
+            $this->removeUnitsAdjustment($item, $promotion);
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getConfigurationFormType()
+    {
+        return 'sylius_promotion_action_percentage_discount_configuration';
+    }
+
+    /**
+     * @param OrderItemInterface $item
+     * @param array $distributedAmounts
+     * @param PromotionInterface $promotion
+     */
+    private function setUnitsAdjustments(OrderItemInterface $item, array $distributedAmounts, PromotionInterface $promotion)
+    {
+        $units = $item->getUnits();
+        $units->first();
+        foreach ($distributedAmounts as $key => $amount) {
+            if (0 === $amount) {
+                continue;
+            }
+
+            $unit = $this->getNextUnit($units);
+            $this->addAdjustmentToUnit($unit, $amount, $promotion);
+        }
+    }
+
+    /**
+     * @param OrderItemUnitInterface $unit
+     * @param $amount
+     * @param PromotionInterface $promotion
+     */
+    private function addAdjustmentToUnit(OrderItemUnitInterface $unit, $amount, PromotionInterface $promotion)
+    {
+        $adjustment = $this->createAdjustment($promotion, AdjustmentInterface::ORDER_ITEM_PROMOTION_ADJUSTMENT);
+        $adjustment->setAmount(-$amount);
+
+        $unit->addAdjustment($adjustment);
+    }
+
+    /**
+     * @param Collection $units
+     *
+     * @return OrderItemUnitInterface
+     */
+    private function getNextUnit(Collection $units)
+    {
+        $unit = $units->current();
+        if (false === $unit) {
+            throw new \InvalidArgumentException('The number of promotion items is greater than number of units.');
+        }
+
+        $units->next();
+
+        return $unit;
+    }
+
+    /**
+     * @param OrderItemInterface $item
+     * @param PromotionInterface $promotion
+     */
+    private function removeUnitsAdjustment(OrderItemInterface $item, PromotionInterface $promotion)
+    {
+        foreach ($item->getUnits() as $unit) {
+            foreach ($unit->getAdjustments(AdjustmentInterface::ORDER_ITEM_PROMOTION_ADJUSTMENT) as $adjustment) {
```

Comment:
Do you think it is possible to make some reasonable extraction, to avoid nested if in a nested foreach?